app/src/main/java/org/dhis2/data/prefs/PreferenceProviderImpl.kt

```diff
@@ -144,4 +146,29 @@ class PreferenceProviderImpl(val context: Context) : PreferenceProvider {
         SecurePreferences.clearAllValues(context)
         sharedPreferences.edit().clear().apply()
     }
+
+    override fun saveGroupingForProgram(programUid: String, shouldGroup: Boolean) {
+        val mapTypeToken = object : TypeToken<MutableMap<String, Boolean>>() {}.type
+        val groupingPrograms =
+            Gson().fromJson<MutableMap<String, Boolean>>(getString("GROUPING", "{}"), mapTypeToken)
+        if(shouldGroup) {
+            groupingPrograms[programUid] = shouldGroup
+        }else{
+            groupingPrograms.remove(programUid)
+        }
+        val updatedGroupingProgramsJson = Gson().toJson(groupingPrograms)
+        setValue("GROUPING", updatedGroupingProgramsJson)
```

Comment:
this should not go in preferences provider. The idea of preferences provider is to be completely abstract with atomic operations like this 
  ` fun removeValue(key: String)
    fun contains(vararg keys: String): Boolean
    fun getString(key: String, default: String? = null): String?
    fun getInt(key: String, default: Int): Int
    fun getLong(key: String, default: Long): Long?
    fun getBoolean(key: String, default: Boolean): Boolean
    fun getFloat(key: String, default: Float): Float?
    fun getSet(key: String, default: Set<String>): Set<String>?`
    
If we need to access data, then we should have a repository for grouping program that connects to the preferences provider